File: Examples/test-suite/csharp/director_shared_ptr_runme.cs
using director_shared_ptrNamespace;
using System;

public class runme
{
  private class director_shared_ptr_MyBarFoo : Foo
  {
    public override string ping()
    {
      return "director_shared_ptr_MyBarFoo.ping()";
    }

    public override string pong()
    {
      return "director_shared_ptr_MyBarFoo.pong();" + ping();
    }

    public override string upcall(FooBar fooBarPtr)
    {
      return "override;" + fooBarPtr.FooBarDo();
    }

    public override Foo makeFoo()
    {
      return new Foo();
    }
  }

  private class director_shared_ptr_MyBarFooDerived : FooDerived
  {
    public override string ping()
    {
      return "director_shared_ptr_MyBarFooDerived.ping()";
    }

    public override string pong()
    {
      return "director_shared_ptr_MyBarFooDerived.pong();" + ping();
    }

    public override string upcall(FooBar fooBarPtr)
    {
      return "overrideDerived;" + fooBarPtr.FooBarDo();
    }

    public override Foo makeFoo()
    {
      return new Foo();
    }
  }

  private static void check(string got, string expected)
  {
    if (got != expected)
      throw new ApplicationException("Failed, got: " + got + " expected: " + expected);
  }

  static void Main()
  {
    FooBar fooBar = new FooBar();

    Foo myBarFoo = new director_shared_ptr_MyBarFoo();
    check(myBarFoo.ping(), "director_shared_ptr_MyBarFoo.ping()");
    check(Foo.callPong(myBarFoo), "director_shared_ptr_MyBarFoo.pong();director_shared_ptr_MyBarFoo.ping()");
    check(Foo.callUpcall(myBarFoo, fooBar), "override;Bar::Foo2::Foo2Bar()");

    Foo myFoo = myBarFoo.makeFoo();
    check(myFoo.pong(), "Foo::pong();Foo::ping()");
    check(Foo.callPong(myFoo), "Foo::pong();Foo::ping()");
    check(myFoo.upcall(new FooBar()), "Bar::Foo2::Foo2Bar()");

    Foo myFoo2 = new Foo().makeFoo();
    check(myFoo2.pong(), "Foo::pong();Foo::ping()");
    check(Foo.callPong(myFoo2), "Foo::pong();Foo::ping()");

    FooDerived myBarFooDerived = new director_shared_ptr_MyBarFooDerived();
    check(myBarFooDerived.ping(), "director_shared_ptr_MyBarFooDerived.ping()");
    check(FooDerived.callPong(myBarFooDerived), "director_shared_ptr_MyBarFooDerived.pong();director_shared_ptr_MyBarFooDerived.ping()");
    check(FooDerived.callUpcall(myBarFooDerived, fooBar), "overrideDerived;Bar::Foo2::Foo2Bar()");

  }
}

File: Examples/test-suite/rust/director_shared_ptr_runme.rs
#![allow(non_snake_case)]

mod director_shared_ptr;

use director_shared_ptr::{Foo, FooBar, FooDerived, FooDerivedDirector, FooDerivedDirectorOwned, FooDirector, FooDirectorOwned};

struct MyBarFoo;

impl FooDirector for MyBarFoo {
    fn ping(&mut self) -> String {
        "director_shared_ptr_MyBarFoo.ping()".to_string()
    }

    fn pong(&mut self) -> String {
        format!("director_shared_ptr_MyBarFoo.pong();{}", self.ping())
    }

    fn upcall(&mut self, fooBarPtr: Option<&mut FooBar>) -> String {
        format!("override;{}", fooBarPtr.expect("fooBarPtr").FooBarDo())
    }

    fn makeFoo(&mut self) -> Foo {
        Foo::new()
    }
}

struct MyBarFooDerived;

impl FooDirector for MyBarFooDerived {
    fn ping(&mut self) -> String {
        "director_shared_ptr_MyBarFooDerived.ping()".to_string()
    }

    fn pong(&mut self) -> String {
        format!("director_shared_ptr_MyBarFooDerived.pong();{}", FooDirector::ping(self))
    }

    fn upcall(&mut self, fooBarPtr: Option<&mut FooBar>) -> String {
        format!("overrideDerived;{}", fooBarPtr.expect("fooBarPtr").FooBarDo())
    }

    fn makeFoo(&mut self) -> Foo {
        Foo::new()
    }
}

impl FooDerivedDirector for MyBarFooDerived {
    fn ping(&mut self) -> String {
        FooDirector::ping(self)
    }

    fn pong(&mut self) -> String {
        FooDirector::pong(self)
    }

    fn upcall(&mut self, fooBarPtr: Option<&mut FooBar>) -> String {
        FooDirector::upcall(self, fooBarPtr)
    }

    fn makeFoo(&mut self) -> Foo {
        FooDirector::makeFoo(self)
    }
}

fn check(got: String, expected: &str) {
    if got != expected {
        panic!("failed, got: {}, expected: {}", got, expected);
    }
}

fn main() {
    let mut foo_bar = FooBar::new();

    let mut my_bar_foo = FooDirectorOwned::from_proxy(Foo::new(), MyBarFoo);
    check(my_bar_foo.ping(), "director_shared_ptr_MyBarFoo.ping()");
    check(
        Foo::callPong(my_bar_foo.as_proxy_mut()),
        "director_shared_ptr_MyBarFoo.pong();director_shared_ptr_MyBarFoo.ping()",
    );
    check(
        Foo::callUpcall(my_bar_foo.as_proxy_mut(), Some(&mut foo_bar)),
        "override;Bar::Foo2::Foo2Bar()",
    );

    let mut my_foo = my_bar_foo.makeFoo();
    check(my_foo.pong(), "Foo::pong();Foo::ping()");
    check(Foo::callPong(&mut my_foo), "Foo::pong();Foo::ping()");
    check(my_foo.upcall(Some(&mut FooBar::new())), "Bar::Foo2::Foo2Bar()");

    let mut my_foo2 = Foo::new().makeFoo();
    check(my_foo2.pong(), "Foo::pong();Foo::ping()");
    check(Foo::callPong(&mut my_foo2), "Foo::pong();Foo::ping()");

    let mut my_bar_foo_derived = FooDerivedDirectorOwned::from_proxy(FooDerived::new(), MyBarFooDerived);
    let mut my_bar_foo_derived_as_foo = my_bar_foo_derived.as_proxy_mut().as_Foo_mut();
    check(my_bar_foo_derived_as_foo.ping(), "director_shared_ptr_MyBarFooDerived.ping()");
    check(
        Foo::callPong(&mut my_bar_foo_derived_as_foo),
        "director_shared_ptr_MyBarFooDerived.pong();director_shared_ptr_MyBarFooDerived.ping()",
    );
    check(
        Foo::callUpcall(&mut my_bar_foo_derived_as_foo, Some(&mut foo_bar)),
        "overrideDerived;Bar::Foo2::Foo2Bar()",
    );
}
